server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java

```diff
@@ -0,0 +1,135 @@
+
+import org.apache.datasketches.frequencies.ErrorType;
+import org.apache.datasketches.frequencies.ItemsSketch;
+import org.apache.datasketches.memory.Memory;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import io.crate.Streamer;
+
+public class MostCommonValuesSketch<T> {
+
+    private static final int MAX_VALUES = 100;
+
+    private final ItemsSketch<T> sketch;
+    private final SketchStreamer<T> streamer;
+
+    public MostCommonValuesSketch(Streamer<T> streamer) {
+        this.streamer = new SketchStreamer<>(streamer);
+        this.sketch = new ItemsSketch<>(256);
+    }
+
+    public MostCommonValuesSketch(Streamer<T> streamer, StreamInput in) throws IOException {
+        this.streamer = new SketchStreamer<>(streamer);
+        byte[] bytes = in.readByteArray();
+        this.sketch = ItemsSketch.getInstance(Memory.wrap(bytes), this.streamer);
+    }
+
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeByteArray(this.sketch.toByteArray(streamer));
+    }
+
+    @SuppressWarnings("unchecked")
+    public MostCommonValuesSketch<T> merge(MostCommonValuesSketch<?> other) {
+        var otherSketch = (ItemsSketch<T>) other.sketch;
+        this.sketch.merge(otherSketch);
+        return this;
+    }
+
+    public void update(T value) {
+        this.sketch.update(value);
+    }
+
+    public MostCommonValues<T> toMostCommonValues(long valueCount, double approxDistinct) {
+        return toMostCommonValues(valueCount, approxDistinct, Function.identity());
+    }
+
+    public <R> MostCommonValues<R> toMostCommonValues(long valueCount, double approxDistinct, Function<T, R> converter) {
+        var rows = this.sketch.getFrequentItems(ErrorType.NO_FALSE_NEGATIVES);
+        int count = Math.min(MAX_VALUES, rows.length);
+        long[] counts = new long[count];
+        List<R> values = new ArrayList<>();
+        for (int i = 0; i < count; i++) {
+            counts[i] = rows[i].getLowerBound();
+            values.add(converter.apply(rows[i].getItem()));
+        }
+        if (approxDistinct < MAX_VALUES) {
+            return new MostCommonValues<>(values, toFreqs(counts, valueCount));
+        }
+        int cutoff = cutoff(this.sketch.getStreamLength(), approxDistinct, counts);
+        values = values.subList(0, cutoff);
+        counts = Arrays.copyOf(counts, cutoff);
+        return new MostCommonValues<>(values, toFreqs(counts, valueCount));
+    }
+
+    private static double[] toFreqs(long[] counts, long valueCount) {
+        double[] freqs = new double[counts.length];
+        for (int i = 0; i < freqs.length; i++) {
+            freqs[i] = (double) counts[i] / valueCount;
+        }
+        return freqs;
+    }
+
+    private static int cutoff(long valueCount, double approxDistinct, long[] counts) {
+        int cutoff = counts.length;
+
+        long sumFreq = 0;
+        for (int i = 0; i < counts.length - 1; i++) {
```

Comment:
is `- 1` correct here?

Reply:
It is not!  Good catch.

Reply:
Oh no, hang on, it is correct.  We are calculating the sum of frequencies *excluding* the last value, to see if the frequency of the last value is low enough in comparison that it can be excluded.  I'll add a comment.

Reply:
:+1 thx!